### PR TITLE
Add task-def and file flags to local down/ps

### PR DESCRIPTION
**Pull request by iamhopaul123**
**Issue #, if available**:

**Description of changes**:
Added two flags "--task-def" and "--file" for "ps" and "down" options for ecs-cli local for filtering.
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**  
- [x] Unit tests passed
- [x] Integration tests passed
- [N/A] Unit tests added for new functionality
- [x] Listed manual checks and their outputs in the comments ([example](https://github.com/aws/amazon-ecs-cli/pull/750#issuecomment-472623042))
- [ ] Add e2e tests for the new flags
- [ ] Link to issue or PR for the integration tests: 

**Documentation**  
- [ ] Contacted our doc writer
- [ ] Updated our README
----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,6 @@ ecs-cli/vendor/pkg
 ecs-cli/.vscode/*
 ecs-cli/.idea
 .idea/*
+*.DS_Store
+*.yml
+Dockerfile
```
Comment (PettitWesley) [inline .gitignore]: Why did you add `Dockerfile` to this?

Comment (iamhopaul123) [reply .gitignore]: Well, I should remove this line.

File: ecs-cli/modules/cli/local/ps_app.go [hunk 2]
```diff
@@ -63,18 +68,39 @@ const (
...
 // If the --json flag is provided, then output the format as JSON instead.
 func Ps(c *cli.Context) {
+	if err := optionsPreCheck(c); err != nil {
+		logrus.Fatalf("Tasks can be either created by local files or remote files")
+	}
 	containers := listContainers(c)
 	displayContainers(c, containers)
 }
 
+func optionsPreCheck(c *cli.Context) error {
```
Comment (PettitWesley) [inline ecs-cli/modules/cli/local/ps_app.go]: I'd recommend renaming this to `validateOptions` or `validateInput` or `validateFlags`

File: ecs-cli/modules/cli/local/ps_app.go [hunk 3]
```diff
@@ -63,18 +68,39 @@ const (
 // If the --all flag is provided, then list all local ECS task containers.
 // If the --json flag is provided, then output the format as JSON instead.
 func Ps(c *cli.Context) {
+	if err := optionsPreCheck(c); err != nil {
+		logrus.Fatalf("Tasks can be either created by local files or remote files")
```
Comment (PettitWesley) [inline ecs-cli/modules/cli/local/ps_app.go]: This log statement is the same as the error text; it would be better to replace it with `logrus.Fatalf(err)`

Comment (PettitWesley) [reply ecs-cli/modules/cli/local/ps_app.go]: Also- should this be a fatal error? I thought a user can run `ps` without either of these flags.

File: ecs-cli/modules/cli/local/ps_app.go (hunk 2 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: ok I will use validateOptions

File: ecs-cli/modules/cli/local/down_app.go [hunk 4]
```diff
@@ -42,8 +42,30 @@ func Down(c *cli.Context) error {
 		network.Teardown(client)
 	}()
 
+	if err := optionsPreCheck(c); err != nil {
+		logrus.Fatalf("Tasks can be either created by local files or remote files")
+	}
+	// if c.Bool(flags.AllFlag) {
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/down_app.go]: Can we delete commented code? In general commented code creates confusion, noone feels comfortable removing it because you lose context on why it was commented in the first place.

File: ecs-cli/modules/cli/local/ps_app.go (hunk 3 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: good catch. thx. But probably should be \code logrus.Fatalf(err.Error())

Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: > Also- should this be a fatal error? I thought a user can run `ps` without either of these flags.

yes they can. This is to prevent them from using both --file and --task-def flags at the same time

Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: > Also- should this be a fatal error? I thought a user can run `ps` without either of these flags.

there will be no error or alert if they do not put any flags. Just show the pwd containers

File: ecs-cli/modules/cli/local/down_app.go [hunk 5]
```diff
@@ -42,8 +42,30 @@ func Down(c *cli.Context) error {
...
+	// 	return downLocalContainersWithFilters()
+	// }
+	// return downComposeLocalContainers()
+
+	if c.String(flags.TaskDefinitionFileFlag) != "" {
+		return downLocalContainersWithFilters(filters.NewArgs(
+			filters.Arg("label", taskDefinitionLabelValue+"="+c.String(flags.TaskDefinitionFileFlag)),
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/down_app.go]: nit: Instead of concatenating strings with the "+" operator, consider using [fmt.Sprintf](https://golang.org/pkg/fmt/#Sprintf). I find it more readable:
```suggestion
			filters.Arg("label", fmt.Sprintf("%s=%s, taskDefinitionLabelType, c.String(flags.TaskDefinitionFileFlag))
```

File: ecs-cli/modules/cli/local/ps_app.go [hunk 6]
```diff
@@ -35,9 +36,13 @@ import (
...
 // Docker object labels associated with containers created with "ecs-cli local".
 const (
-	// taskDefinitionLabelKey represents the value of the option used to
-	// transform a task definition to a compose file e.g. file path, arn, family.
-	taskDefinitionLabelKey = "ecsLocalTaskDefinition"
+	// taskDefinitionLabelType represents the type of option used to
+	// transform a task definition to a compose file e.g. remoteFile, localFile.
+	// taskDefinitionLabelValue represents the value of the option
+	// e.g. file path, arn, family.
+	taskDefinitionLabelKey   = "ecsLocalTaskDefinition"
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/ps_app.go]: Can we delete this now?

File: ecs-cli/modules/cli/local/down_app.go [hunk 7]
```diff
@@ -42,8 +42,30 @@ func Down(c *cli.Context) error {
...
+			filters.Arg("label", taskDefinitionLabelType+"="+"localFile"),
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/down_app.go]: Can we move `remoteFile` and `localFile` to constants? Ideally, these should be defined in `create_app.go`. We have a note to fix these duplications here: https://github.com/aws/amazon-ecs-cli/projects/2#card-22758290

File: ecs-cli/modules/cli/local/down_app.go [hunk 8]
```diff
@@ -65,19 +87,17 @@ func downComposeLocalContainers() error {
...
-		Filters: filters.NewArgs(
-			filters.Arg("label", taskDefinitionLabelKey),
-		),
-		All: true,
+		Filters: args,
+		All:     true,
 	})
 	if err != nil {
-		logrus.Fatalf("Failed to list containers with label=%s due to %v", taskDefinitionLabelKey, err)
+		logrus.Fatalf("Failed to list containers with label=%s due to %v", taskDefinitionLabelValue, err)
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/down_app.go]: Should we list all `args` here instead of just the `taskDefinitionLabelValue` ?
```suggestion
		logrus.Fatalf("Failed to list containers with filters %v due to %v", args, err)
```

File: ecs-cli/modules/cli/local/ps_app.go (hunk 3 above)
Comment (PettitWesley) [reply ecs-cli/modules/cli/local/ps_app.go]: Ah I see. In that case, how about something like this for the error: 
```
fmt.Errorf("%s and %s can not be used together", flags.TaskDefinitionTaskFlag, flags.TaskDefinitionFileFlag)
```

File: .gitignore [hunk 9]
```diff
@@ -8,3 +8,5 @@ ecs-cli/vendor/pkg
 ecs-cli/.vscode/*
 ecs-cli/.idea
 .idea/*
+*.DS_Store
+*.yml
```
Comment (efekarakus) [inline .gitignore]: We don't want to ignore YAML files, we have things like `buildspec.yml` that should be commited. Can you please remove this?

File: ecs-cli/modules/cli/local/ps_app.go [hunk 10]
```diff
@@ -63,18 +68,39 @@ const (
 // If the --all flag is provided, then list all local ECS task containers.
 // If the --json flag is provided, then output the format as JSON instead.
 func Ps(c *cli.Context) {
+	if err := validateOptions(c); err != nil {
+		logrus.Fatalf(err.Error())
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/ps_app.go]: `logrus.Fatal(err.Error())` without the `f`

File: ecs-cli/modules/cli/local/ps_app.go [hunk 11]
```diff
@@ -63,18 +68,39 @@ const (
...
+	}
 	containers := listContainers(c)
 	displayContainers(c, containers)
 }
 
+func validateOptions(c *cli.Context) error {
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/ps_app.go]: nit: can you move this fn to `create_app.go`? It seems like it should belong there.

File: ecs-cli/modules/cli/local/down_app.go [hunk 12]
```diff
@@ -42,8 +42,30 @@ func Down(c *cli.Context) error {
 		network.Teardown(client)
 	}()
 
+	if err := validateOptions(c); err != nil {
+		logrus.Fatalf("Tasks can be either created by local files or remote files")
```
Comment (efekarakus) [inline ecs-cli/modules/cli/local/down_app.go]: Should this also be `logrus.Fatal(err.Error())`?

File: ecs-cli/modules/cli/local/ps_app.go (hunk 3 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: Alright, I will fix that

File: ecs-cli/modules/commands/local/local_command.go [hunk 13]
```diff
@@ -60,12 +60,20 @@ func upCommand() cli.Command {
...
 		Action: local.Down,
 		Flags: []cli.Flag{
 			cli.BoolFlag{
 				Name:  flags.AllFlag,
-				Usage: "Stop and remove all running local ECS tasks.",
+				Usage: "Stops and removes all running containers",
+			},
+			cli.StringFlag{
+				Name:  flags.TaskDefinitionTaskFlag,
+				Usage: "Stops and removes all running containers matching the task family or ARN",
```
Comment (efekarakus) [inline ecs-cli/modules/commands/local/local_command.go]: Nice!

File: ecs-cli/modules/commands/local/local_command.go [hunk 14]
```diff
@@ -60,12 +60,20 @@ func upCommand() cli.Command {
...
+			},
+			cli.StringFlag{
+				Name:  flags.TaskDefinitionFileFlag,
+				Usage: "Stops and removes all running containers matching the path",
```
Comment (efekarakus) [inline ecs-cli/modules/commands/local/local_command.go]: What do you think about?
```suggestion
				Usage: "Stops and removes all running containers matching the task definition file path",
```

File: ecs-cli/modules/cli/local/down_app.go (hunk 4 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/down_app.go]: Oh my bad. I forgot to delete them

File: ecs-cli/modules/cli/local/down_app.go (hunk 5 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/down_app.go]: ok thx. Fixed

File: ecs-cli/modules/cli/local/ps_app.go (hunk 6 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: Yes we can!

File: ecs-cli/modules/cli/local/down_app.go (hunk 7 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/down_app.go]: Oh that is also what I was thinking about. I was about to do that. Probably tomorrow.

File: ecs-cli/modules/cli/local/down_app.go (hunk 8 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/down_app.go]: Yeah I agree with you.

File: .gitignore (hunk 9 above)
Comment (iamhopaul123) [reply .gitignore]: ok

File: ecs-cli/modules/cli/local/down_app.go (hunk 12 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/down_app.go]: Yeah I forgot to update here

File: ecs-cli/modules/commands/local/local_command.go (hunk 14 above)
Comment (iamhopaul123) [reply ecs-cli/modules/commands/local/local_command.go]: Sure

File: ecs-cli/modules/cli/local/ps_app.go (hunk 11 above)
Comment (iamhopaul123) [reply ecs-cli/modules/cli/local/ps_app.go]: fixed

**Comment by efekarakus** [comment]
Looks great! @iamhopaul123 can you also run the integration tests to validate that nothing breaks?

File: ecs-cli/integ/cmd/compose.go [hunk 15]
```diff
@@ -85,6 +85,7 @@ func TestServiceUp(t *testing.T, p *Project) {
 		"up",
 		"--cluster-config",
 		p.ConfigName,
+		"--create-log-groups",
```
Comment (hencrice) [inline ecs-cli/integ/cmd/compose.go]: nit: Just a general comment for best practices. I normally try to limit the number of changes in a single PR such that one PR represents a self-contained, logical set of changes. For this PR, that means creating a new PR for the bug fixes, merge it in, then continue working on this PR but keep it strictly for the feature you are working on.

By doing so, it has the following benefits:

1. In case we accidentally introduce any bug in the feature once it's merged, we can just revert the feature without reverting the bug fixes
2. If an issue starts showing up after a certain PR is merged, then it's easier to pinpoint the issue.

**Comment by iamhopaul123** [comment]
## Manual Testing

#### Configuration

##### Parameters

```yaml
# docker-compose.yml
version: "3"
networks:
  ecs-local-network:
    external: true
services:
  app:
    build:
      context: .
    ports:
    - "6001:80"
    networks: 
    - ecs-local-network
    environment:
      AWS_DEFAULT_REGION: "us-east-1"
      AWS_CONTAINER_CREDENTIALS_RELATIVE_URI: "/creds"
      ECS_CONTAINER_METADATA_URI: "http://169.254.170.2/v3"
    labels:
      ecsLocalTaskDefType: 'remoteFile'
      ecsLocalTaskDefVal: 'arn:aws:ecs:us-east-1:685593908319:task-definition/efe-test:1'
```

```yaml
# docker-compose.yml
version: "3"
networks:
  ecs-local-network:
    external: true
services:
  app:
    build:
      context: .
      dockerfile: "Dockerfile1"
    ports:
    - "6002:70"
    networks: 
    - ecs-local-network
    environment:
      AWS_DEFAULT_REGION: "us-east-1"
      AWS_CONTAINER_CREDENTIALS_RELATIVE_URI: "/creds"
      ECS_CONTAINER_METADATA_URI: "http://169.254.170.2/v2"
    labels:
      ecsLocalTaskDefType: 'localFile'
      ecsLocalTaskDefVal: '/a/b/c'
```

```yaml
# docker-compose.yml
version: "3"
networks:
  ecs-local-network:
    external: true
services:
  app:
    build:
      context: .
    ports:
    - "6004:30"
    networks: 
    - ecs-local-network
    environment:
      AWS_DEFAULT_REGION: "us-east-1"
      AWS_CONTAINER_CREDENTIALS_RELATIVE_URI: "/creds"
      ECS_CONTAINER_METADATA_URI: "http://169.254.170.2/v3"
    labels:
      ecsLocalTaskDefType: 'remoteFile'
      ecsLocalTaskDefVal: 'arn:aws:ecs:us-east-1:685593908319:task-definition/efe-test:3'
```

```Dockerfile
#Dockerfile
FROM ubuntu:latest
RUN apt-get update \
 && apt-get -y install curl \
 && apt-get -y install awscli \
 && apt-get -y install jq \
 && apt-get -y install net-tools
CMD /bin/bash -c 'while true; do sleep 30; done;'
```

Note that, I put three "docker-compose.yml" in different directories to avoid recreation.

#####  Environment Setup

`make build`
`ecs-cli local up`
`docker-compose -f docker-compose.local.yml up -d`

#### Tests

##### Lists all running containers matching the task family or ARN

`ecs-cli local ps --task-def arn:aws:ecs:us-east-1:685593908319:task-defini
tion/efe-test:3
`

```
CONTAINER ID        IMAGE                 STATUS              PORTS                  NAMES                    TASKDEFINITION
ae907267d51a        untitledfolder2_app   Up 4 minutes        0.0.0.0:6004->30/tcp   /untitledfolder2_app_1   arn:aws:ecs:us-east-1:685593908319:task-definition/efe-test:3
```

##### Lists all running containers matching the task definition file path

`ecs-cli local ps --file /a/b/c`

```
CONTAINER ID        IMAGE                STATUS              PORTS                  NAMES                   TASKDEFINITION
f1ae19996920        untitledfolder_app   Up 5 minutes        0.0.0.0:6002->70/tcp   /untitledfolder_app_1   /a/b/c
```

##### List all running containers matching the task definition file path and task family or ARN

`ecs-cli local ps --file /a/b/c --task-def arn:aws:ecs:us-east-1:6855939083
19:task-definition/efe-test:1`

```
FATA[0000] task-def and file can not be used together
```

##### Stops and removes all running containers matching the task definition file path

`ecs-cli local down --file /a/b/c`

```
INFO[0000] Stop and remove 1 container(s)               
INFO[0010] Stopped container with id f1ae19996920       
INFO[0010] Removed container with id f1ae19996920       
INFO[0010] 2 other task container(s) running locally, skipping network removal 
```

##### Stops and removes all running containers matching the task family or ARN

`ecs-cli local down --task-def arn:aws:ecs:us-east-1:685593908319:task-defi
nition/efe-test:1`

```
INFO[0000] Stop and remove 1 container(s)               
INFO[0010] Stopped container with id 7b5b1b82ffc3       
INFO[0010] Removed container with id 7b5b1b82ffc3       
INFO[0010] 1 other task container(s) running locally, skipping network removal 
```

File: ecs-cli/integ/cmd/compose.go (hunk 15 above)
Comment (iamhopaul123) [reply ecs-cli/integ/cmd/compose.go]: Good advice! Next time when I encounter problems of another fields, I will create a new branch a submit another PR. Thanks for the advice and explanation.

File: ecs-cli/modules/cli/local/create_app.go [hunk 16]
```diff
@@ -18,12 +18,39 @@ package local
...
+	taskDefinitionLabelValue = "ecsLocalTaskDefVal"
+)
+
+const (
+	localTaskDefType  = "localFile"
+	remoteTaskDefType = "remoteFile"
+)
+
+const (
+	ecsLocalDockerComposeFileName = "docker-compose.local.yml"
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/local/create_app.go]: I have a different constant name for this: `LocalOutDefaultFileName ` https://github.com/aws/amazon-ecs-cli/pull/785/files#diff-c72fc3b09ef311d6734cb6f6b9c9b9a6R36.

File: ecs-cli/modules/cli/local/create_app.go [hunk 17]
```diff
@@ -18,12 +18,39 @@ package local
...
+	// taskDefinitionLabelValue represents the value of the option
+	// e.g. file path, arn, family.
+	taskDefinitionLabelType  = "ecsLocalTaskDefType"
+	taskDefinitionLabelValue = "ecsLocalTaskDefVal"
+)
+
+const (
+	localTaskDefType  = "localFile"
+	remoteTaskDefType = "remoteFile"
+)
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/local/create_app.go]: I'm also a little confused why these four constants are defined in `create_app.go` right now, as I don't see them being used in the create function. Perhaps it would make sense to move these constants to a different file still under the `local` package namespace?

File: ecs-cli/modules/cli/local/create_app.go [hunk 18]
```diff
@@ -18,12 +18,39 @@ package local
...
+
 func Create(c *cli.Context) {
 	// 1. read in task def (from file or arn)
 	// 2. parse task def into go object
 	// 3. write to docker-compose.local.yml file
 	fmt.Println("foo") // placeholder
 }
+
+func validateOptions(c *cli.Context) error {
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/local/create_app.go]: Similarly, not sure why this function lives in this file?

File: ecs-cli/modules/cli/local/create_app.go (hunk 17 above)
Comment (efekarakus) [reply ecs-cli/modules/cli/local/create_app.go]: I think the idea is that eventually `Create()` will eventually add these as Docker labels to the containers so that we know if the containers were started with a `localFile` or `remoteFile`.

File: ecs-cli/modules/cli/local/create_app.go (hunk 18 above)
Comment (efekarakus) [reply ecs-cli/modules/cli/local/create_app.go]: The flags that are being validated apply to all the sub-commands (`create`, `up`, `down`, `ps`). Should this live in some other file? 🤔 I think we decided to move it in `create_app.go` because it's the first command that users will use for local but I can see how that's not a strong reason.
